### ISSUE TITLE
Failing Build: Function 'machs' at (1) has no IMPLICIT type

### ISSUE THREAD

**Issue by crollinson** [EDmodel/ED2#210]
I'm fast forwarding a lot to the latest version of ED on GitHub and I'm unable to build.  I'm getting a few variables about missing type.  It looks like this goes back to #200 by @manfredo89 since when I add some calls back in, it gets further.  (see commit cb30c21985f440a45503373c76b66a9b1e4ccf11 "major revision..."

I'll keep debugging since it looks like this isn't the only place with issues, but I want to see if I'm the only one that has had this fail... I see the travis build has been failing lately. 

Note for @manfredo89, it's very hard to see if this is where things came in because the commit that I think this links back to had  555 changed files and too many differences for the git diff or "blame" functions to work.  In the future, please make these smaller commits to make them easier to track.  

Also, please make sure all compile checks from a clean build work before pulling into mainstream since I'd be surprised if I'm the only one getting this error.

**Comment by manfredo89** [comment]
Hi @crollinson, I would suggest you try to incorporate the next pull requests (#205 #206 ) and try to recompile. I've recently been in contact with @mpaiao about merging these two small updates and I think it should soon be feasible. I'm saying this because I've modified the way install.sh calls the make (this is also the reason for Travis failing to build).
However this should not be the cause of missing types.
I am aware of a bug in file ed_read_ed10_20.f90:907 where it should now be
`         call init_ed_site_vars(cpoly)
`
instead of
`         call init_ed_site_vars(cpoly,cgrid%lat(ipy))`
though this was not producing errors with the --kind A compilation on intel platform (with the flags we were using).
If the error you are getting is different please post the code, I will have a look at it.
Also regarding the size of the PR there were a couple of issues. Some files appear to have changed even though they were not. This was due partly to automatic change of permissions. I will try in the future to be more efficient but I think it would really help if there were some guidelines on the git workflow to update the code.

**Comment by manfredo89** [comment]
Hi @crollinson did you manage to compile?
Also be aware that the optimization entry in the ED2IN was removed so you have to delete the following lines

```

   !---------------------------------------------------------------------------------------!
   ! IOPTINPT -- Optimization configuration. (Currently not used)                          !
   !---------------------------------------------------------------------------------------!
   NL%IOPTINPT = ''
   !---------------------------------------------------------------------------------------!

```

**Comment by crollinson** [comment]
I haven't not made progress on compiling ED (out sick and then have other priorities at the moment).

However, adding in the couple missing types allowed me to make progress.  See the last couple commits here: https://github.com/crollinson/ED2/commits/master

I was able to get through state_vars, but now am getting the same types of errors about missing types in fuse_fiss_utils.  e.g. 
```
fuse_fiss_utils.f90:1215.30:

      cpatch%lint_shv(recc) = qslif8(can_prss,cpatch%leaf_temp(recc))
                              1
Error: Function 'qslif8' at (1) has no IMPLICIT type
```

Compiling ED is *not* my priority at the moment.  I will submit a PR with the fixes I find when I get a chance, but it could be a while.

**Comment by serbinsh** [comment]
@crollinson what compiler flags are you using in your opt file?

**Comment by mpaiao** [comment]
@crollinson I downloaded the mainline and @manfredo89's latest PR (which I haven't merged yet), and this is what I see in line 1215:
```fortran
      cpatch%lint_shv(recc) = qslif(can_prss,cpatch%leaf_temp(recc))
```
In any case, this line should call qslif instead of qslif8, because all variables in this line are single precision.   Could it be that this came from merging your branch with the mainline?

**Comment by crollinson** [comment]
@mpaiao @manfredo89  I just pulled the ED2 mainline into a clean clone with the pull requests you just pulled in and am still getting the same errors reported in in the original thread, so it's not from the merge with my personal fork.

@serbinsh my make script can be found here: https://github.com/crollinson/ED2/blob/master/ED/build/make/include.mk.opt.ulmus 

This is I think identical to what worked in January to compile my PalEON branch on my server here.  It's a variant of the linux compile with MPI turned on... maybe that's why Marcos & Manfredo haven't been getting these errors?

**Comment by crollinson** [comment]
After fixing deleted variable calls in a ton of files, I'm able to get ED to compile.  When I get a chance, I'll try pulling in the newest PR and see if I can still compile.

To see full set of changes that had to be made, go here: https://github.com/crollinson/ED2/commits/master

I'll merge them into a cleaner commit set of commits for the PR

**Comment by manfredo89** [comment]
Hi @crollinson, I made a couple of comments to your commits. I think the optimiz_coms should not be reintroduced unless there is some implementation (it was just an empty file and an empty entry in the ED2IN). Check carefully for the therm_lib / therm_lib8 because there were some mismatch.
Regarding MPI variables like ierr I think it's ok to reintroduce them. The compiler was throwing a warning of unused variables but they are indeed referenced. Not sure why this is happening, maybe they are sort of mpi intrinsics so they need not be defined.

**Comment by crollinson** [comment]
@manfredo89 thanks for taking a look, but everything I re-introduced was necessary for the compile to work... I think there was one with the therm_lib where things got a little confused, but double precision wouldn't work for my compile, so I had/have to reset the functions to single precision.

**Comment by mpaiao** [comment]
Files with F90 extension (as opposed to f90) are files with pre-processing instructions.  Most of these instructions refer to MPI or HDF5 libraries.  What probably happened is that the variable was declared outside the pre-compiler if statement, but used only inside the pre-compiler if statement.  When you turn off the MPI option, the compiler will warn the unused variable, but deleting it will make the compilation crash when the variable is turned on.  

The ideal solution is this case is to add the variable declaration inside the pre-compiler instructions.  For example:
```fortran
#if defined(RAMS_MPI)
   integer :: ierr
#endif
```

I'm having trouble installing MPI libraries in the clusters I am using, so I'm only able to test the model with the MPI option turned off...

**Comment by manfredo89** [comment]
Ok, I indeed suspected this was related to the preprocessor however I did not know about this difference with the extension, that is useful to know. On my cluster I managed to compile and run with Intel MPI so that's why I was saying that these variables could be kind of intrinsic or implicit.

**Comment by crollinson** [comment]
Pull request submitted; see #211

This issue can be closed once #211 is pulled in
